### PR TITLE
mds/server: fix rare race when waitting for osdmap

### PR DESCRIPTION
**Pull request by xiexingguo**
If wait_for_map() returns true, we have got the requested osdmap
at the specified epoch, so we shall try again to do further
verification whether we have the specific pool or not.
    
The above case can happen because we drop the objecter internal
rwlock during the switch between the objecter->with_osdmap()
and objecter->wait_for_map() methods.
    
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

File: src/mds/Server.cc [hunk 1]
```diff
@@ -4264,10 +4286,14 @@ void Server::handle_set_vxattr(MDRequestRef& mdr, CInode *cur,
...
-      respond_to_request(mdr, r);
-      return;
+
+      if (r < 0) { // check again, result code could have changed
+        if (r == -ENOENT)
+          r = -EINVAL;
+        respond_to_request(mdr, r);
+        return;
+      }
     }
```
Comment (ukernel) [inline src/mds/Server.cc]: lots of duplicated code here, maybe we move them into seperate function

File: src/mds/Server.cc [hunk 2]
```diff
@@ -4501,7 +4501,7 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
   set<SimpleLock*> rdlocks, wrlocks, xlocks;
   file_layout_t *dir_layout = NULL;
   CInode *cur;
-  if (name == "ceph.dir.layout")
+  if (name.compare(0, 15, "ceph.dir.layout") == 0)
```
Comment (ukernel) [inline src/mds/Server.cc]: But inside handle_remove_vxattr(), it still checks name == "ceph.dir.layout"

**Comment by xiexingguo** [comment]
@ukernel Thank you for your careful review.

I have done upgrades based on your gorgeous advice, could you please have a look again?

**Comment by xiexingguo** [comment]
retest this please

**Comment by jcsp** [comment]
@ukernel could you review the updated code please?

**Comment by xiexingguo** [comment]
Tested at:
http://www.daisycloud.org:9091/xxg-2016-06-23_20:47:59-fs-wip-xxg-testing-fs-2016-06-23---basic-plana/

**Comment by ukernel** [comment]
please remove commit "server: handle attribute name in a more graceful way".  handle_remove_vxattr only support removing whole dir.layout

**Comment by xiexingguo** [comment]
@ukernel Done. Thanks!

**Comment by xiexingguo** [comment]
@ukernel Thanks for the merge, zheng:-)

**Pull request merged**